### ISSUE TITLE
Download tool not recognizing already downloaded (cached) versions

### ISSUE THREAD

**Issue by jacobmsft** [microsoft/OSSGadget#50]
```
E:\test>...\oss-defog\bin\Debug\netcoreapp3.1\oss-defog.exe pkg:npm/postcss@7.0.27

E:\test>dir
...

 Directory of E:\test

05/08/2020  08:34 PM    <DIR>          .
05/08/2020  08:34 PM    <DIR>          ..
05/08/2020  08:34 PM    <DIR>          npm-postcss@7.0.27

E:\test>...\oss-defog\bin\Debug\netcoreapp3.1\oss-defog.exe pkg:npm/postcss@7.0.27

E:\test>dir

 Directory of E:\test

05/08/2020  08:34 PM    <DIR>          .
05/08/2020  08:34 PM    <DIR>          ..
05/08/2020  08:34 PM    <DIR>          npm-postcss@7.0.27
05/08/2020  08:34 PM    <DIR>          npm-postcss@7.0.27-637245668633217700
```

**Comment by scovetta** [comment]
This is by design, since we can't tell in advance whether the bits in the `npm-postcss@7.0.27` directory are the same as the bits we would download, without downloading.

I suppose we could have a cache marker/metadata file, but I don't think this would be a common use case. 

Alternatively, we could fail if the directory exists, or take a `--force` option, but I think this might add complexity for not a lot of benefit.

I'll leave this issue open in case you/others have other thoughts.

**Comment by jacobmsft** [comment]
That makes sense w.r.t integrity, but in a controlled environment, it shouldn't matter, IMO. The requirement that I have is to run multiple OSSGadget tools on the same package - so for defogger, it downloads the package once, for characteristics, it downloads again, etc - which wastes bandwidth and processing power. It might make sense to add a --download-path, and assume that the download folder, if present, is a valid cache.

**Comment by scovetta** [comment]
That's a good point.

What if we added an option to most of the tools to take in a path to work on? Something like this (from oss-detect-cryptography):

```csharp
if (target.StartsWith("pkg:", StringComparison.InvariantCulture))
{
  var purl = new PackageURL(target);
  results = await detectCryptographyTool.AnalyzePackage(purl);
}
else if (Directory.Exists(target))
{
   results = await detectCryptographyTool.AnalyzeDirectory(target);
}
else if (File.Exists(target))
...
```

We should do it in a more reusable way, but would something along those lines work? We couldn't do it for oss-health or oss-find-source, but for the analyzers I think this would make sense.

**Comment by jacobmsft** [comment]
Yes, that sounds good to me. Not all tools need to work on a package - it can work on a given folder too - oss-detect-cryptography, for example need not even have the notion of packages - it just detects cryptography content in files. Stays close to the "do one well and do it well" philosophy.

**Comment by gfs** [comment]
@jacobmsft was this fixed in #59 ?

**Comment by jacobmsft** [comment]
Yes, this is done, the issue can be closed.
